src/Microsoft.Health.Dicom.Blob/Features/Storage/BlobFileStore.cs

```diff
@@ -172,9 +183,43 @@ public async Task DeleteFileIfExistsAsync(long version, string partitionName, Ca
         BlockBlobClient blobClient = GetInstanceBlockBlobClient(version, partitionName);
         _logger.LogInformation("Trying to delete DICOM instance file with watermark '{Version}'.", version);
 
+        EmitTelemetry("DeleteFileIfExistsAsync", OperationType.Input);
+
         await ExecuteAsync(() => blobClient.DeleteIfExistsAsync(DeleteSnapshotsOption.IncludeSnapshots, conditions: null, cancellationToken));
     }
 
+    private static readonly Action<ILogger, string, bool, Exception> LogBlobClientOperationDelegate =
+        LoggerMessage.Define<string, bool>(
+            LogLevel.Information,
+            default,
+            "Operation '{OperationName}' processed. Using external store {IsExternalStore}.");
+
+    private static readonly Action<ILogger, string, long, Exception> LogBlobClientOperationWithStreamDelegate =
+        LoggerMessage.Define<string, long>(
+            LogLevel.Information,
+            default,
+            "Operation '{OperationName}' processed stream length '{StreamLength}'.");
+
+    private void EmitTelemetry(string operationName, OperationType operationType, long? streamLength = null)
+    {
+        _blobFileStoreMeter.BlobFileStoreOperationCount.Add(
+            1,
+            BlobFileStoreMeter.BlobFileStoreOperationTelemetryDimension(operationName, operationType, _blobClient.IsExternal));
+
+        if (streamLength == null)
+        {
+            LogBlobClientOperationDelegate(_logger, operationName, _blobClient.IsExternal, null);
+        }
+        else
+        {
+            var lenght = streamLength.Value;
+            LogBlobClientOperationWithStreamDelegate(_logger, operationName, streamLength.Value, null);
+            _blobFileStoreMeter.BlobFileStoreOperationStreamSize.Add(
+                lenght,
+                BlobFileStoreMeter.BlobFileStoreOperationTelemetryDimension(operationName, operationType, _blobClient.IsExternal));
```

Comment:
This API is not the most appealing. Is there a way to statically type this?

Reply:
This could potentially be very out-of-scope.

If there is no such API, and only key-value pairs are accepted, then we can put some thought into later coming up with some sort of nifty type-safe API like:

```csharp
_blobFileStoreMeter.BlobFileStoreOperationStreamSize.Add(
    length,
    operationName,
    operationType,
    _blobClient.IsExternal);

// OR

_blobFileStoreMeter.BlobFileStoreOperationStreamSize.Add<OperationContext>(
    length,
    new OperationContext(operationName, operationType, _blobClient.IsExternal);

// OR
using OperationContext context = new(operationName, operationType, _blobClient.IsExternal);

_blobFileStoreMeter.BlobFileStoreOperationStreamSize.Add(length, context);
```

Reply:
There's actually something in R9 about this, so we'll get to it later